lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/RandomSpec.hs

```diff
@@ -0,0 +1,154 @@
+    ( Depth (..), DerivationType (..), Index (..), Passphrase (..), XPrv )
+import Cardano.Wallet.Primitive.AddressDerivation.Common
+    ( Key (..) )
+import Cardano.Wallet.Primitive.AddressDerivation.Random
+    ( deriveAccountPrivateKey
+    , deriveAddressPrivateKey
+    , encodeDerivationPath
+    , generateKeyFromSeed
+    , minSeedLengthBytes
+    )
+import Cardano.Wallet.Primitive.AddressDerivationSpec
+    ()
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( IsOurs (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Random
+    ( RndState (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Address (..) )
+import Crypto.Hash
+    ( hash )
+import Crypto.Hash.Algorithms
+    ( Blake2b_224, SHA3_256 )
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Base58
+    ( bitcoinAlphabet, encodeBase58 )
+import Data.Digest.CRC32
+    ( crc32 )
+import Test.Hspec
+    ( Spec, describe, it )
+import Test.QuickCheck
+    ( Arbitrary (..), Property, choose, property, vectorOf, (.&&.), (===) )
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Write as CBOR
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as BS
+
+spec :: Spec
+spec = do
+    describe "Random Address Discovery Properties" $ do
+
+        it "goldenAddress gives never isOurs for arbitrary generated root private key" $ do
+            property prop_goldenAddressIsOurs
+        it "isOurs works as expected during key derivation" $ do
+            property prop_keyDerivationObeysIsOurs
+
+
+{-------------------------------------------------------------------------------
+                               Properties
+-------------------------------------------------------------------------------}
+
+prop_goldenAddressIsOurs
+    :: Key 'RootK XPrv
+    -> Property
+prop_goldenAddressIsOurs rk =
+    let
+        goldenAddr = "DdzFFzCqrht2qSNuod2j3HQdxQYu7ehMnHqPMK6ZCZc1oTBfFJFTaqMF62rzWsJWZhbrN15uA4Bsp6M7t5WkqfumdnjLjZ5xRk8szuCd"
+    in
+        isOurs (Address goldenAddr) (RndState rk) === (False, RndState rk)
+
+prop_keyDerivationObeysIsOurs
+    :: Passphrase "seed"
+    -> Passphrase "encryption"
+    -> Index 'Hardened 'AccountK
+    -> Index 'Soft 'AddressK
+    -> Key 'RootK XPrv
+    -> Property
+prop_keyDerivationObeysIsOurs seed encPwd accIx addrIx rk' =
+    isOurs address (RndState rootXPrv) === (True, RndState rootXPrv) .&&.
+    isOurs address (RndState rk') === (False, RndState rk')
+  where
+    rootXPrv@(Key rootXPrv') = generateKeyFromSeed seed encPwd
+    rootXPub = Key (toXPub rootXPrv')
+    accXPrv = deriveAccountPrivateKey encPwd rootXPrv accIx
+    (Key addrXPrv) = deriveAddressPrivateKey encPwd accXPrv addrIx
+    addrXPub = toXPub addrXPrv
+
+    encodeAddressAttributes = mempty
+        <> CBOR.encodeMapLen 1
+        <> CBOR.encodeWord8 1
+        <> encodeDerivationPath rootXPub accIx addrIx
+
+    address = Address $ encodeBase58 bitcoinAlphabet
```

Comment:
this is one important addition : to encode to base58, the rest is from Matthias wallet-prototype